barretenberg/cpp/src/barretenberg/stdlib/primitives/group/cycle_scalar.cpp

```diff
@@ -82,126 +100,122 @@ template <typename Builder> cycle_scalar<Builder> cycle_scalar<Builder>::create_
-        // Step 4: apply range constraints to validate both slices represent the expected contributions to *this.lo and
-        // *this,hi
-        limb_1_lo.create_range_constraint(lo_bits_in_limb_1);
-        limb_1_hi.create_range_constraint(hi_bits_in_limb_1);
-
-        // construct *this.lo out of:
-        // a. `limb0` (the first NUM_LIMB_BITS bits of scalar)
-        // b. `limb_1_lo` (the first LO_BITS - NUM_LIMB_BITS) of limb1
-        lo = limb0 + (limb_1_lo * BigScalarField::shift_1);
-
-        const uint256_t limb_2_shift = uint256_t(1) << (BigScalarField::NUM_LIMB_BITS - lo_bits_in_limb_1);
-        const uint256_t limb_3_shift =
-            uint256_t(1) << ((BigScalarField::NUM_LIMB_BITS - lo_bits_in_limb_1) + BigScalarField::NUM_LIMB_BITS);
-
-        // construct *this.hi out of limb2, limb3 and the remaining term from limb1 not contributing to `lo`
-        hi = limb_1_hi.add_two(limb2 * limb_2_shift, limb3 * limb_3_shift);
+        lo.set_origin_tag(scalar.get_origin_tag());
+        hi.set_origin_tag(scalar.get_origin_tag());
+        return;
+    }
+
+    // Step 1: Ensure the bigfield scalar fits into LO_BITS + HI_BITS by reducing if necessary.
+    // Note: we can tolerate the scalar being > ScalarField::modulus, because performing a scalar mul implicitly
+    // performs a modular reduction.
+    if (scalar.get_maximum_value() >= (uint512_t(1) << (LO_BITS + HI_BITS))) {
+        scalar.self_reduce();
     }
+
+    field_t limb0 = scalar.binary_basis_limbs[0].element;
+    field_t limb1 = scalar.binary_basis_limbs[1].element;
+    field_t limb2 = scalar.binary_basis_limbs[2].element;
+    field_t limb3 = scalar.binary_basis_limbs[3].element;
+
+    uint256_t limb1_max = scalar.binary_basis_limbs[1].maximum_value;
+
+    // Step 2: Ensure that limb0 only contains at most NUM_LIMB_BITS. If not, slice off the excess and add it into limb1
+    if (scalar.binary_basis_limbs[0].maximum_value > BigScalarField::DEFAULT_MAXIMUM_LIMB) {
+        const uint256_t limb0_value = limb0.get_value();
+        const uint256_t limb0_lo_value = limb0_value.slice(0, BigScalarField::NUM_LIMB_BITS);
+        const uint256_t limb0_hi_value = limb0_value >> BigScalarField::NUM_LIMB_BITS;
+        field_t limb0_lo = field_t::from_witness(ctx, limb0_lo_value);
+        field_t limb0_hi = field_t::from_witness(ctx, limb0_hi_value);
+
+        // Constrain the limb0 decomposition to be well formed
+        uint256_t limb0_hi_max = (scalar.binary_basis_limbs[0].maximum_value >> BigScalarField::NUM_LIMB_BITS);
+        const uint64_t limb0_hi_bits = limb0_hi_max.get_msb() + 1;
+        limb0_lo.create_range_constraint(BigScalarField::NUM_LIMB_BITS);
+        limb0_hi.create_range_constraint(static_cast<size_t>(limb0_hi_bits));
+        limb0.assert_equal(limb0_lo + (limb0_hi * BigScalarField::shift_1));
+
+        // Move the excess bits from limb0 into limb1
+        limb1 += limb0_hi;
+        limb1_max += limb0_hi_max;
+        limb0 = limb0_lo;
```

Comment:
Don't think it really matters, but we should set the max value of limb 0 to `DEFAULT_MAXIMUM_LIMB`

Reply:
So `limb0` here is a copy of the limb within the bigfield element, so changes to it here do not effect the bigfield element that we passed in. For that reason I'm wondering if it would be more confusing to than not to update the max value